### PR TITLE
PE-463: download folder

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

**Comment by matibat** [comment]
TODO: update readme

File: README.md [hunk 1]
```diff
@@ -720,6 +721,26 @@ Users can perform a bulk upload by using the upload-file command on a target fol
...
 
+### Downloading a Folder with Files<a id="download-folder"></a>
+
+You can download a folder on chain with the `download-folder` command.
+
+```shell
+ardrive download-folder -f "47f5bde9-61ba-49c7-b409-1aa4a9e250f6"
+```
+
+By specifying the `--dest-folder-path` you can choose the parent folder where to download the folder on chain. When the parameter is missing, it defaults to the current working directory (`./`).
```
Comment (matibat) [inline README.md:732]:
```suggestion
By specifying the `--local-path` you can choose the parent folder where to download the folder on chain. When the parameter is missing, it defaults to the current working directory (`./`).
```

File: README.md [hunk 2]
```diff
@@ -720,6 +721,26 @@ Users can perform a bulk upload by using the upload-file command on a target fol
...
+
+```shell
+ardrive download-folder -f "47f5bde9-61ba-49c7-b409-1aa4a9e250f6" --dest-folder-path /my_ardrive_downloads/
```
Comment (matibat) [inline README.md:735]:
```suggestion
ardrive download-folder -f "47f5bde9-61ba-49c7-b409-1aa4a9e250f6" --local-path /my_ardrive_downloads/
```

File: package.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
 	},
 	"types": "./lib/index.d.ts",
 	"dependencies": {
-		"ardrive-core-js": "1.4.0",
+		"ardrive-core-js": "1.5.0-alpha-0",
```
Comment (matibat) [inline package.json:11]: TODO: update

File: README.md [hunk 4]
```diff
@@ -720,6 +722,26 @@ Users can perform a bulk upload by using the upload-file command on a target fol
 ardrive upload-file --local-path /path/to/folder  --parent-folder-id "9af694f6-4cfc-4eee-88a8-1b02704760c0" -w /path/to/wallet.json
 ```
 
+### Downloading a Folder with Files<a id="download-folder"></a>
+
+You can download a folder on chain with the `download-folder` command.
```
Comment (arielmelendez) [inline README.md:727]:
```suggestion
You can download a folder from ArDrive to your local machine with the `download-folder` command.
```

File: README.md [hunk 5]
```diff
@@ -720,6 +722,44 @@ Users can perform a bulk upload by using the upload-file command on a target fol
...
+### Downloading a Folder with Files<a id="download-folder"></a>
+
+You can download a folder from ArDrive to your local machine with the `download-folder` command. In the following examples, assume that a folder with ID "47f5bde9-61ba-49c7-b409-1aa4a9e250f6" exists in your drive and is named "MyArDriveFolder".
+
+```shell
+# Downloads "MyArDriveFolder" into the current working directory, i.e. ./MyArDriveFolder/
+ardrive download-folder -f "47f5bde9-61ba-49c7-b409-1aa4a9e250f6"
+```
+
+By specifying the `--local-path` option, you can choose the local parent folder into which you the on-chain folder will be downloaded. When the parameter is omitted, its value defaults to the current working directory (i.e. `./`).
```
Comment (arielmelendez) [inline README.md:734]:
```suggestion
By specifying the `--local-path` option, you can choose the local parent folder into which the on-chain folder will be downloaded. When the parameter is omitted, its value defaults to the current working directory (i.e. `./`).
```

File: src/commands/download_folder.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		const maxDepth = await parameters.getMaxDepth(Number.MAX_SAFE_INTEGER);
+		const destOutputPath = parameters.getParameterValue(DownloadLocalPathParameter) || '.';
+		const [destFolderPath, customFolderName] = getOutputFolderPathAndName(destOutputPath);
+		let outputPath: string;
+
+		if (await parameters.getIsPrivate()) {
+			const wallet = await parameters.getRequiredWallet();
+			const ardrive = cliArDriveFactory({
+				wallet,
+				feeMultiple: parameters.getOptionalBoostSetting()
```
Comment (arielmelendez) [inline src/commands/download_folder.ts:38]: Might be able to omit this.

File: src/commands/index.ts [hunk 7]
```diff
@@ -25,3 +25,4 @@ import './send_ar';
 import './send_tx';
 import './tx_status';
 import './upload_file';
+import './download_folder';
```
Comment (arielmelendez) [inline src/commands/index.ts:28]: Please preserve alphabetic order here.

**Review by arielmelendez — DISMISSED** [review]

File: README.md [hunk 8]
```diff
@@ -101,10 +101,12 @@ ardrive upload-file --wallet-file /path/to/my/wallet.json --parent-folder-id "f0
...
-        4. [Uploading Multiple Files](#multi-file-upload)
-        5. [Fetching the Metadata of a File Entity](#fetching-the-metadata-of-a-file-entity)
-        6. [Uploading Manifests](#uploading-manifests)
-        7. [Hosting a Webpage with Manifest](#hosting-a-webpage-with-manifest)
+        4. [Downloading a Folder with Files](#download-folder)
+        5. [Uploading Multiple Files](#multi-file-upload)
+        6. [Fetching the Metadata of a File Entity](#fetching-the-metadata-of-a-file-entity)
+        7. [Uploading Manifests](#uploading-manifests)
+        8. [Hosting a Webpage with Manifest](#hosting-a-webpage-with-manifest)
+        9. [Create New Drive and Upload Folder Pipeline Example](#create-upload-pipeline)
```
Comment (fedellen) [inline README.md:109]: This section `create-upload-pipeline` was removed on dev due to incompatibility with drive bundles. Probably showing up here from a merge conflict.

Comment (matibat) [reply README.md]: It was because a wrong merge conflict resolution on my side. It's removed at e5bb0b9353f696fc3b526bf066a433c107967e70

**Review by fedellen — APPROVED** [review]

**Review by arielmelendez — APPROVED** [review]

**Review by fedellen — DISMISSED** [review]

**Review by agsuy — APPROVED** [review]
Tests just got added. Approved.

File: bats_test/downloads/folder-downloads/cleanup.sh [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+DIR="$(cd "$(dirname "${BATS_TEST_FILENAME}")" >/dev/null 2>&1 && pwd)"
+
+rm -rf \
+    "${DIR}/Files With a Variety of Extensions Public" \
```
Comment (arielmelendez) [inline bats_test/downloads/folder-downloads/cleanup.sh:6]: Why not just `${DIR}/*`?

Comment (matibat) [reply bats_test/downloads/folder-downloads/cleanup.sh:6]: At `${DIR}/` there is the `.bats` test files

**Review by fedellen — DISMISSED** [review]

**Review by arielmelendez — DISMISSED** [review]

**Review by fedellen — APPROVED** [review]

**Pull request merged**